src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs

```diff
@@ -20,60 +21,72 @@ internal static class GlobalJsonReader
         /// </summary>
         /// <returns>A <see cref="Dictionary{String,String}"/> of MSBuild SDK versions from a global.json if found, otherwise <code>null</code>.</returns>
-        public static Dictionary<string, string> GetMSBuildSdkVersions(SdkResolverContext context)
+        public static Dictionary<string, string> GetMSBuildSdkVersions(SdkResolverContext context, out bool wasGlobalJsonRead, string fileName = GlobalJsonFileName)
         {
-            if (string.IsNullOrWhiteSpace(context?.ProjectFilePath))
+            bool wasRead = false;
+
+            wasGlobalJsonRead = wasRead;
+
+            if (string.IsNullOrWhiteSpace(context?.ProjectFilePath) || string.IsNullOrWhiteSpace(fileName))
             {
                 // If the ProjectFilePath is not set, an in-memory project is being evaluated and there's no way to know which directory to start looking for a global.json
                 return null;
             }
 
             DirectoryInfo projectDirectory = Directory.GetParent(context.ProjectFilePath);
 
-            if (!TryGetPathOfFileAbove(GlobalJsonFileName, projectDirectory, out FileInfo globalJsonPath))
+            if (projectDirectory == null || !TryGetPathOfFileAbove(fileName, projectDirectory, out FileInfo globalJsonPath))
             {
                 return null;
             }
 
-            string contents = File.ReadAllText(globalJsonPath.FullName);
-
-            // Look ahead in the contents to see if there is an msbuild-sdks section.  Deserializing the file requires us to load
-            // Newtonsoft.Json which is 500 KB while a global.json is usually ~100 bytes of text.
-            if (contents.IndexOf(MSBuildSdksPropertyName, StringComparison.Ordinal) == -1)
+            if (!globalJsonPath.Exists)
             {
-                return null;
-            }
+                FileCache.TryRemove(globalJsonPath, out var _);
 
-            try
-            {
-                return Deserialize(contents);
-            }
-            catch (Exception e)
-            {
-                // Failed to parse "{0}". {1}
-                var message = string.Format(CultureInfo.CurrentCulture, Strings.FailedToParseGlobalJson, globalJsonPath, e.Message);
-                context.Logger.LogMessage(message);
                 return null;
             }
-        }
 
-        /// <summary>
-        /// Deserializes a global.json and returns the MSBuild SDK versions
-        /// </summary>
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static Dictionary<string, string> Deserialize(string value)
-        {
-            return JsonConvert.DeserializeObject<GlobalJsonFile>(value).MSBuildSdks;
-        }
+            // Add a new file to the cache if it doesn't exist.  If the file is already in the cache, read it again if the file has changed
+            (DateTime LastWriteTime, Lazy<Dictionary<string, string>> Lazy) result = FileCache.AddOrUpdate(
```

Comment:
Could we use remove func here? Because previously removing func for concurrent dictionary ( https://github.com/NuGet/NuGet.Client/pull/4095/files ) improved perf, I guess this is very similar situation.

Reply:
Also do we really need `Lazy` here? Because we always end up evaluating the value.

Reply:
Great question.  Yes the `Lazy<T>` is necessary because of how `ConcurrentDictionary<TKey, TValue>.AddOrUpdate()` works:

[From the docs: ](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-6.0#System_Collections_Concurrent_ConcurrentDictionary_2_AddOrUpdate__0__1_System_Func__0__1__1__)

> For modifications and write operations to the dictionary, ConcurrentDictionary<TKey,TValue> uses fine-grained locking to ensure thread safety. (Read operations on the dictionary are performed in a lock-free manner.) The addValueFactory and updateValueFactory delegates may be executed multiple times to verify the value was added or updated as expected. However, they are called outside the locks to avoid the problems that can arise from executing unknown code under a lock. Therefore, AddOrUpdate is not atomic with regards to all other operations on the ConcurrentDictionary<TKey,TValue> class.

This means that two threads will only add one item, but end up calling the factory method if they're both called at the same time.  To ensure that the factory is only called once, you wrap it in a `Lazy<T>`.  The `Lazy` object can be created multiple times, but all of the calls receive the object that was created first.  All threads then access `Value` which is thread-safe so all threads wait for the factory to actually run (if it hasn't already) or they receive the cached value.

One blog post that explains more is here: https://andrewlock.net/making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/

I've been bitten by this in the past which is the only reason I know about this limitation: https://github.com/dotnet/msbuild/pull/4797